chapter3.adoc

```diff
@@ -1,71 +1,63 @@
+This chapter outlines the security enhancements defined for the Debug Module as non-ISA extension. The debug operations listed below are modified by the non-ISA extension. All features in this chapter must be implemented in Debug Module to achieve external debug security. If any hart in the system implements the Sdec extension, the Debug Module must also implement the non-ISA extension.  
 
-The affected debug operations, including:
+* Halt
+* Reset 
+* Keepalive 
+* Abstract commands (Access Register, Quick Access, Access Memory)
+* System bus access
 
-    - Halt request
-    - Reset request 
-    - Keepalive request 
-    - Abstract commands 
+=== External Debug Security Extensions Discovery 
 
-Will be addressed as follows.
+The ISA and non-ISA external debug security extensions impose security constraints and introduce non-backward-compatible changes. The presence of the extensions can be determined by polling the `allsecured` or/and `anysecured` bits in dmstatus <<regdmstatus>>.  If the field `allsecured` or `anysecured` is set to 1, it represents that all or any selected harts adopt the Sdsec extension. When any hart adopts the Sdsec extension, it indicates the Debug Module implements Debug Module Security Extension as described in this chapter.
 
-=== Debug Module Security Extension Discovery 
+=== Halt 
 
-The Debug Module Security Extension imposes security constraints and introduces non-backward-compatible changes. The presence of the Debug Module Security Extension can be determined by polling the `allsecured/anysecured` bits in dmstatus.
-
-=== Halt
-
-The hart can only be halted when it is running in privilege levels specified in <<dbgpriv>>. Otherwise, the halt request will remain pending. Consequently, the response to the external debugger request may take longer than one second, and this constraint must be eliminated.
-
-When machine mode is not permitted (mdbgen[i] set to 0) to engage in debugging, the halt-on-reset (resethaltreq) operation must fail and raise security fault error. The debugger could check the error by polling `allsecfault/anysecfault` bits in dmstatus as specified in <<regdmstatus>>. 
+The halt behavior for a hart is detailed in <<sdsecextdbg>>. According to _The RISC-V Debug Specification_ cite:[dbgspec],  a halt request must be responded within one second. However, this constraint must be eliminated as the request might be pending due to the situations where debugging is disallowed. Additionally, when M-mode is not permitted (mdbgen[i] set to 0) to engage in debugging, the halt-on-reset (resethaltreq) operation must fail and raise security fault error. The debugger could check the error by polling `allsecfault` or/and `anysecfault` fields in dmstatus for selected harts, as specified in <<regdmstatus>>. 
```

Comment:
I think the halt-on-reset operation is to debug the very first instruction executed. If we make the halt pending and let it enter debug mode when the privilege changes to S-mode, the operation just debugs the first debug-allowed instruction. I think we should inform the debugger that the very first instruction is not debuggable instead of stop at an unwanted spot.

Reply:
Sorry my memory might be bad but I thought we discussed this in one TG meeting and decided to stop at first instruction of debuggable mode for halt-on-reset?

Reply:
Discussed in last week's TG meeting and agreed to update the spec, halt-on-reset will be pending and halt the core eventually after it enters S-mode.

Reply:
Sounds good, I agree that seems simplest.